### PR TITLE
Test sink name

### PR DESCRIPTION
**Pull request by Naireen**
Add tests to check the name of the transform

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java [hunk 1]
```diff
@@ -477,6 +480,39 @@ public void testBuildWrite() {
     assertThat(write.getComputeNumShards(), equalTo(originalSharding));
   }
 
+  @Test
+  public void testWriteTransformName() {
```
Comment (y1chi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java:484]: Anywhere in runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow we can put these tests? It is very specific for dataflow that I don't feel like it is good idea to put it under sdk.

Also can we create a github issue as a TODO for the tests for a more generic solution for identifying these transforms when submitting to dataflow via v1b3 API. We should remove these tests after we have a more safe solution.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java [hunk 2]
```diff
@@ -477,6 +480,39 @@ public void testBuildWrite() {
...
+    assertThat(write.getComputeNumShards(), nullValue());
+
+    pipeline
+        .apply(Create.timestamped(inputs, timestamps).withCoder(StringUtf8Coder.of()))
+        .setIsBoundedInternal(IsBounded.UNBOUNDED)
+        .apply(Window.into(FixedWindows.of(Duration.millis(1))))
+        .apply(write);
+    // This should not change, this ensures that the transform name is what the dataflow runner
+    // expects.
+    logged.verifyDebug("WriteAutoShardedBundlesToTempFiles");
```
Comment (y1chi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java:513]: Can we test more directly? For example we may be able to get the step name from translated job specification https://github.com/apache/beam/blob/d329a7e9b85d729a474395abfe1975c0f3649550/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L199
just in case the name is somehow changed or removed after translation

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java [hunk 3]
```diff
@@ -1293,6 +1295,10 @@ private void storageWrite(boolean autoSharding) throws Exception {
     p.apply(source).apply("WriteToBQ", write);
     p.run().waitUntilFinish();
 
+    // This should not change, this ensures that the transform name is what the dataflow runner
+    // expects.
+    logged.verifyDebug("/StorageApiLoads");
```
Comment (y1chi) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java:1300]: ditto.

Comment (Naireen) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java (hunk 2 above)
Comment (Naireen) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteFilesTest.java]: Done.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 4]
```diff
@@ -229,6 +238,126 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
     return options;
   }
 
+  // Test that the transform names are what we expect them to be
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:241]: Can we make the comment more clear what is expected and why. Also please add a TODO comment with github issue link for more stable solution.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 5]
```diff
@@ -229,6 +238,126 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
...
+  public void testStorageWriteApiTransformNames() throws IOException, Exception {
+
+    DataflowPipelineOptions options = buildPipelineOptions();
+    options.setRunner(DataflowRunner.class);
+    Pipeline p = Pipeline.create(options);
+
+    p.traverseTopologically(new RecordingPipelineVisitor());
+    SdkComponents sdkComponents = createSdkComponents(options);
+
+    List<String> inputs = Lists.newArrayList();
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:252]: you could probably just inline some input elements with Create.of(), since the input is not important we can make it much simpler.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 6]
```diff
@@ -229,6 +238,126 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
...
+                    .setDatasetId("dataset")
+                    .setTableId("table"))
+            .withSchema(
+                new TableSchema()
+                    .setFields(
+                        new ArrayList<>(
+                            ImmutableList.of(
+                                new TableFieldSchema().setName("description").setType("STRING")))));
+
+    p.apply(Create.of(inputs).withCoder(StringUtf8Coder.of()))
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:277]: can we combine these two tests (StorageApiLoads and WriteAutoSharededBundlesToTempFiles) into 1. We can just use different branches.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 7]
```diff
@@ -229,6 +238,126 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
...
+            PipelineOptionsFactory.as(DataflowPipelineOptions.class));
+
+    JobSpecification jobSpecification =
+        t.translate(
+            p,
+            pipelineProto,
+            sdkComponents,
+            DataflowRunner.fromOptions(options),
+            Collections.emptyList());
+    assertAllStepOutputsHaveUniqueIds(jobSpecification.getJob());
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:293]: Is this check relevant to what we are testing?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 8]
```diff
@@ -229,6 +238,126 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
...
+
+    // Assert that at least one of the steps added was the StorageApiLoads step. This is added to
+    // ensure the name doesn't change.
+    boolean foundStep = false;
+    for (Step step : jobSpecification.getJob().getSteps()) {
+      if (getString(step.getProperties(), PropertyNames.USER_NAME).contains("/StorageApiLoads")) {
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:299]: I think StorageWriteApi should also be included in this step name? We can be more specific that StorageWriteApi step contains "/StorageApiLoads"

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java (hunk 4 above)
Comment (Naireen) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: Updated, plus added link to issue.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java (hunk 8 above)
Comment (Naireen) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: Done.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java (hunk 7 above)
Comment (Naireen) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: No, so I removed it.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java (hunk 6 above)
Comment (Naireen) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: Hm, since these are unit tests, and using different pipeline shapes, I think it would be better to keep them as separate tests (though there will be admittedly some overlap in the setup code.)

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java (hunk 5 above)
Comment (Naireen) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: Done

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 9]
```diff
@@ -229,6 +237,115 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
     return options;
   }
 
+  // Test that the transform names for Storage Write API for streaming pipelines are what we expect
+  // them to be. This is required since the Windmill backend expects the step to contain that name.
+  // For a more stable solution, we use use urn, but that is not currently used in the legacy java
+  // worker.
+  // TODO:(https://github.com/apache/beam/issues/29338) Pass in URN information to GCP Runner.
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:244]:
```suggestion
  // TODO:(https://github.com/apache/beam/issues/29338) Pass in URN information to Dataflow Runner.
```

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 10]
```diff
@@ -229,6 +237,115 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
     return options;
   }
 
+  // Test that the transform names for Storage Write API for streaming pipelines are what we expect
+  // them to be. This is required since the Windmill backend expects the step to contain that name.
+  // For a more stable solution, we use use urn, but that is not currently used in the legacy java
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:242]:
```suggestion
  // For a more stable solution, we should use URN, but that is not currently used in the legacy java
```

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 11]
```diff
@@ -229,6 +237,115 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
...
+            sdkComponents,
+            DataflowRunner.fromOptions(options),
+            Collections.emptyList());
+
+    // Assert that at least one of the steps added was the autosharded write. This is added to
+    // ensure the name doesn't change.
+    boolean foundStep = false;
+    for (Step step : jobSpecification.getJob().getSteps()) {
+      if (getString(step.getProperties(), PropertyNames.USER_NAME)
+          .contains("/WriteAutoShardedBundlesToTempFiles")) {
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:342]: nit:WriteMyFile/WriteAutoShardedBundlesToTempFiles?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 12]
```diff
@@ -229,6 +237,115 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
...
+      }
+    }
+    assertTrue(foundStep);
+  }
+
+  // Test that the transform names added for TextIO writes with autosharding. This is required since
+  // the Windmill backend expects the file write autosharded step to contain that name.
+  // For a more stable solution, we use use urn, but that is not currently used in the legacy java
+  // worker.
+  // TODO:(https://github.com/apache/beam/issues/29338) Pass in URN information to GCP Runner.
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:306]:
```suggestion
  // TODO:(https://github.com/apache/beam/issues/29338) Pass in URN information to Dataflow Runner.
```

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 13]
```diff
@@ -229,6 +237,115 @@ private static DataflowPipelineOptions buildPipelineOptions() throws IOException
...
+          .contains("StorageWriteApi/StorageApiLoads")) {
+        foundStep = true;
+      }
+    }
+    assertTrue(foundStep);
+  }
+
+  // Test that the transform names added for TextIO writes with autosharding. This is required since
+  // the Windmill backend expects the file write autosharded step to contain that name.
+  // For a more stable solution, we use use urn, but that is not currently used in the legacy java
```
Comment (y1chi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:304]:
```suggestion
  // For a more stable solution, we should use URN, but that is not currently used in the legacy java
```

**Review by y1chi — APPROVED** [review]
LGTM

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java (hunk 11 above)
Comment (Naireen) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: So its WriteMyFile/WriteFiles/FinalizeTempFileBundles/, I changed this to check for "WriteFiles/WriteAutoShardedBundlesToTempFile, since we don't care about what the user has set.
